### PR TITLE
Add `onFailedAttempt` option

### LINKED ISSUE #3: Add `onReattempt` option
it would be nice to have an onError function hooked up to https://github.com/tim-kos/node-retry#retryoperationattemptfn-timeoutops

### PR DESCRIPTION
**Pull request by acostalima**
Resolves #3

### CODE REVIEW

**Comment by acostalima** [comment]
ping @sindresorhus

**Comment by acostalima** [comment]
ping @sindresorhus

File: index.js [hunk 1]
```diff
@@ -20,10 +20,15 @@ class AbortError extends Error {
 }
 
 module.exports = (input, opts) => new Promise((resolve, reject) => {
+	opts = Object.assign({
+		onFailedAttempt: () => {},
+		retries: 10
```
Comment (evocateur) [inline index.js]: This is already the default of `retry`, why do we need bookkeeping here? If a user cares about "attempts remaining", presumably they're capable of setting an explicit value for `retries` and doing the math themselves.

File: index.js [hunk 2]
```diff
@@ -32,7 +37,9 @@ module.exports = (input, opts) => new Promise((resolve, reject) => {
 				} else if (err instanceof TypeError) {
 					operation.stop();
 					reject(err);
-				} else if (!operation.retry(err)) {
+				} else if (operation.retry(err)) {
+					opts.onFailedAttempt(err, attemptNo, attemptsLeft);
```
Comment (evocateur) [inline index.js]: Decorating `err.attempt = attempt` and calling `opts.onFailedAttempt(err)` feels like a more natural API for node callbacks.

File: test.js [hunk 3]
```diff
@@ -8,10 +8,10 @@ const fixtureErr = new Error('fixture');
 test('retries', async t => {
 	let i = 0;
 
-	const ret = await m(async attempts => {
+	const ret = await m(async attemptNo => {
```
Comment (evocateur) [inline test.js]: Changing `attempts` to `attemptNo` seems rather arbitrary.

Comment (acostalima) [reply test.js]: This change was deliberate: `attempts` is actually the current attempt number, according to `retry`'s docs, and not the total number of attempts as the variable name implies. It was not obvious to me when I first start developing this feature, that is why I made the change to be more clear.

File: index.js (hunk 2 above)
Comment (acostalima) [reply index.js]: Agreed.

File: index.js (hunk 1 above)
Comment (acostalima) [reply index.js]: I understand your point, however the proposal for this feature discussed in #5 includes the computation of the number of attempts left. Concerning the default for `retries`, I'm aware that `retry` already sets it. This is just a safeguard to ensure the math for `attemptsLeft` remains correct even if the default value for `retries` changes in the future for whatever reason.

**Comment by acostalima** [comment]
ping @sindresorhus

File: readme.md [hunk 4]
```diff
@@ -52,6 +52,12 @@ Type: `Object`
 
 Options are passed to the [`retry`](https://github.com/tim-kos/node-retry#retryoperationoptions) module.
 
+##### onFailedAttempt(err, attemptNo, attemptsLeft)
+
+Type: `Function`
+
+Callback invoked on each retry. Receives the error thrown by `input` as the first argument, the current attempt number as the second argument and the number of attempts left as the third argument.
```
Comment (sindresorhus) [inline readme.md]: This needs to be updated to reflect the code (That it's no properties on the error).

I'd also like to see the docs here describe the use-case for this option, as commented in https://github.com/sindresorhus/p-retry/issues/3#issuecomment-344953574.

File: index.js [hunk 5]
```diff
@@ -32,7 +37,11 @@ module.exports = (input, opts) => new Promise((resolve, reject) => {
 				} else if (err instanceof TypeError) {
 					operation.stop();
 					reject(err);
-				} else if (!operation.retry(err)) {
+				} else if (operation.retry(err)) {
+					err.attemptNo = attemptNo;
```
Comment (sindresorhus) [inline index.js]: It should be named `attemptNumber`, not `attemptNo`. Applies to everywhere `attemptNo` is currently used.

**Comment by sindresorhus** [comment]
@acostalima Sorry for the long the delay. I just had too many PRs to review.

**Comment by acostalima** [comment]
@sindresorhus no worries 👌 
MR updated according to your latest feedback. Thanks!

**Comment by sindresorhus** [comment]
Looks great now 🙌 Thanks for your patience, @acostalima :)

**Comment by acostalima** [comment]
@sindresorhus no problem. Thanks! :D

**Pull request merged**